### PR TITLE
Make date helper an interface

### PR DESCRIPTION
**Pull request by leahtaylor**
Make accounts dates helper an interface
Add implementation of dates helper
Update unit tests

Add new method - Get LocalDateFromDateTimeString

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -18,6 +18,7 @@
     <maven-compiler-plugin.version>3.5.1</maven-compiler-plugin.version>
     <maven-surefire-plugin.version>2.21.0</maven-surefire-plugin.version>
     <structured-logging.version>1.2.0-rc1</structured-logging.version>
+    <spring-framework.version>5.0.8.RELEASE</spring-framework.version>
```
Comment (Mathsh) [inline pom.xml]: Are we using the correct version? cos everywhere else we declare the use of spring like so, 

See, company-accounts-api and document-generator 

<parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.0.3.RELEASE</version>
        <relativePath/> 
    </parent>

       <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-hateoas</artifactId>
            <scope>compile</scope>
        </dependency>

File: src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java [hunk 2]
```diff
@@ -1,26 +1,16 @@
...
 
-public class AccountsDates {
-
-    private static final String PERIOD_START = "periodStart";
-    private static final String PERIOD_END = "periodEnd";
-    private static final String DATE_FORMAT_YYYYMMDD = "yyyy-MM-dd";
-    private static final String DATE_FORMAT_D_MMMM_YYYY = "d MMMM yyyy";
+import org.springframework.stereotype.Service;
 
+@Service
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java]: Should this be at interface level?

File: src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    /**
+     *Calculate balance sheet dates display format depending on range between period start and end dates
+     * 
+     * 
+     * @param periodStart accounting period start date
+     * @param periodEnd accounting period end date 
+     * @param isSameYear
+     * @return
+     * 
+     *
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java]: excess '*' can remove some

File: src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            nextDate = currentDate.plusDays(i);
+            if (nextDate.isAfter(now)) {
+                break;
+            }
+            futureDays.add(nextDate);
+        }
+
+        return futureDays;
+    }
+
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java]: can remove the gap

File: src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java [hunk 5]
```diff
@@ -29,34 +19,34 @@
...
      * @return
      */
-    public String convertDateToString(LocalDate date) {
-        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_FORMAT_YYYYMMDD);
-        return date.format(formatter);
-    }
+    public String convertDateToString(LocalDate date);
 
     /**
      * Formats a {@link String} date from format '2017-03-05' to '3 March 2005'
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java]: The formats are implementation specific. Can you add the example (from format '2017-03-05' to '3 March 2005') to the implementation instead please?

File: src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java [hunk 6]
```diff
@@ -65,25 +55,7 @@ public String convertLocalDateToDisplayDate(LocalDate date) {
...
-
-        String date = localDate.format(dateFormatDesired);
-        String time = localDateTime.format(timeFormatDesired).toLowerCase();
-
-        timeObject.put("date", date);
-        timeObject.put("time", time);
-
-        return timeObject;
-    }
+    public Map<String, String> getDateAndTime(String dateString);
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java]: The JavaDoc is implementation specific details. A different implementation might have different time and date formats. Can you include "in the format 'D_MMMM_YYYY' 'h:mm a' " in the implementation JavaDoc instead please

File: src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    /**
+     * Returns a Java 8 {@link LocalDate} from given date/time string
+     * 
+     * @param dateTimeString
+     * @return
+     * 
+     */
+    @Override
+    public LocalDate getLocalDatefromDateTimeString(String dateTimeString) {
+        Instant instant = Instant.parse(dateTimeString);
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java]: Will this work with the dateTimeString in any date & time format?

File: src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java (hunk 2 above)
Comment (ChrisJBurns) [reply src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java]: its normally on impl level

File: src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java (hunk 6 above)
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java]: Resolved in 054c90f

File: src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java (hunk 5 above)
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/accountsDates/AccountsDates.java]: Resolved in 054c90f

File: src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java (hunk 3 above)
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java]: Resolved in 054c90f

File: src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java (hunk 4 above)
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java]: Resolved in 054c90f

File: src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    /**
+     * Takes a {@link String} in "yyyy-MM-dd'T'HH:mm:ss.SSSZ" format and converts it to a Java 8 {@link LocalDate}
+     * 
+     * @param dateTimeString
+     * @return 
+     */
+    @Override
+    public LocalDate getLocalDatefromDateTimeString(String dateTimeString) {
+        Instant instant = Instant.parse(dateTimeString);
+        LocalDateTime localDateTime = LocalDateTime.ofInstant(instant, ZoneId.of(ZoneOffset.UTC.getId()));
```
Comment (jsimons1989) [inline src/main/java/uk/gov/companieshouse/accountsDates/impl/AccountsDatesImpl.java]: As this is a library, can you make the time zone a parameter please? This is assuming all times passed into this helper are in UTC timezone. A library shouldn't make assumptions.

File: pom.xml [hunk 9]
```diff
@@ -64,6 +72,22 @@
...
+      <artifactId>spring-boot-starter-data-mongodb</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-actuator</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-hateoas</artifactId>
+    </dependency>
```
Comment (paulforsyth) [inline pom.xml]: are all these needed?

**Pull request merged**